packages/plugin-ext/src/main/browser/view/view-registry.ts

```diff
@@ -30,42 +36,51 @@ export class ViewRegistry {
-        const widget = new ViewsContainerWidget(viewContainer);
-        this.containersWidgets.set(viewContainer.id, widget);
-        this.updatePromiseChain(() => {
-            if (widget && !this.applicationShell.getTabBarFor(widget)) {
-                this.applicationShell.addWidget(widget, {
-                    area: ApplicationShell.isSideArea(location) ? location : 'left'
-                });
-            }
-        });
+    @postConstruct()
+    init() {
+        this.addPermissionPromise = this.applicationStateService.reachedState(READY);
     }
 
-    registerView(location: string, view: View) {
-        const widget = this.containersWidgets.get(location);
-        if (widget) {
-            widget.addView(view);
+    registerViewContainer(location: string, viewsContainer: ViewContainer, containerViews: View[]): void {
+        if (this.containerWidgets.has(viewsContainer.id)) {
+            return;
         }
-    }
+        const containerWidget = new ViewsContainerWidget(viewsContainer, containerViews);
+        this.containerWidgets.set(viewsContainer.id, containerWidget);
 
-    onRegisterTreeView(viewId: string, treeViewWidget: TreeViewWidget) {
-        this.containersWidgets.forEach(async (viewsContainerWidget: ViewsContainerWidget) => {
-            await this.promiseChain;
-            if (viewsContainerWidget.hasView(viewId)) {
-                viewsContainerWidget.addWidget(viewId, treeViewWidget);
-                this.applicationShell.activateWidget(viewsContainerWidget.id);
+        // add to the promise chain
+        this.addPermissionPromise = this.addPermissionPromise.then(() => {
+            if (this.applicationShell.getTabBarFor(containerWidget)) {
+                return;
             }
+            this.applicationShell.addWidget(containerWidget, {
+                area: ApplicationShell.isSideArea(location) ? location : DEFAULT_LOCATION
+            });
+
+            // update container
+            this.treeViewWidgets.forEach((treeViewWidget: TreeViewWidget, viewId: string) => {
+                if (containerWidget.hasView(viewId) && !containerWidget.hasWidget(treeViewWidget)) {
+                    containerWidget.addWidget(viewId, treeViewWidget);
+                }
+            });
         });
     }
 
-    private updatePromiseChain(callback: () => void): void {
-        if (!this.promiseChain) {
-            this.promiseChain = this.applicationStateService.reachedState('ready');
+    registerTreeView(viewId: string, treeViewWidget: TreeViewWidget): void {
+        this.treeViewWidgets.set(viewId, treeViewWidget);
+
+        if (this.applicationStateService.state !== READY) {
+            return;
         }
-        this.promiseChain = this.promiseChain.then(callback);
+        // update containers
+        this.containerWidgets.forEach((containerWidget: ViewsContainerWidget) => {
+            if (containerWidget.hasView(viewId) && !containerWidget.hasWidget(treeViewWidget)) {
```

Comment:
could these 3 lines be extracted in 1 method ? `updateContainerViews()` ? (and method reused above line 66)

Reply:
not exactly the same
```
            this.treeViewWidgets.forEach((treeViewWidget: TreeViewWidget, viewId: string) => {
                if (containerWidget.hasView(viewId) && !containerWidget.hasWidget(treeViewWidget)) {
                    containerWidget.addWidget(viewId, treeViewWidget);
                }
            });
```
and

```
        this.containerWidgets.forEach((containerWidget: ViewsContainerWidget) => {
            if (containerWidget.hasView(viewId) && !containerWidget.hasWidget(treeViewWidget)) {
                containerWidget.addWidget(viewId, treeViewWidget);
            }
        });
```

Reply:
```
addTreeViewWidget
```
done